group_theory/free_abelian_group.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2018 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+
+Free abelian groups as abelianizatin of free groups.
+-/
+
+import group_theory.free_group
+import group_theory.abelianization
+
+universes u v
+
+variables (α : Type u)
+
+def free_abelian_group : Type u :=
+additive $ abelianization $ free_group α
+
+instance : add_comm_group (free_abelian_group α) :=
+@additive.add_comm_group _ $ abelianization.comm_group _
+
+variable {α}
+
+namespace free_abelian_group
+
+def of (x : α) : free_abelian_group α :=
+abelianization.of $ free_group.of x
+
+instance : has_coe α (free_abelian_group α) :=
+⟨of⟩
+
+theorem coe_def (x : α) : (x : free_abelian_group α) = of x :=
+rfl
+
+section to_comm_group
+
+variables {β : Type v} [add_comm_group β] (f : α → β)
+
+def to_add_comm_group (x : free_abelian_group α) : β :=
+@abelianization.to_comm_group _ _ (multiplicative β) _ (@free_group.to_group _ (multiplicative β) _ f) _ x
+
+def to_add_comm_group.is_add_group_hom :
+  is_add_group_hom (to_add_comm_group f) :=
+⟨λ x y, @is_group_hom.mul _ (multiplicative β) _ _ _ (abelianization.to_comm_group.is_group_hom _) x y⟩
+
+local attribute [instance] to_add_comm_group.is_add_group_hom
+
+@[simp] lemma to_add_comm_group.add (x y : free_abelian_group α) :
+  to_add_comm_group f (x + y)
+  = to_add_comm_group f x +
+    to_add_comm_group f y :=
+is_add_group_hom.add _ _ _
+
+@[simp] lemma to_add_comm_group.neg (x : free_abelian_group α) :
+  to_add_comm_group f (-x)
+  = -to_add_comm_group f x :=
+is_add_group_hom.neg _ _
+
+@[simp] lemma to_add_comm_group.zero :
+  to_add_comm_group f 0 = 0 :=
+is_add_group_hom.zero _
+
+@[simp] lemma to_add_comm_group.of (x : α) :
+  to_add_comm_group f (of x) = f x :=
+by unfold of; unfold to_add_comm_group; simp
+
+theorem to_add_comm_group.unique
+  (g : free_abelian_group α → β) [is_add_group_hom g]
+  (hg : ∀ x, g (of x) = f x) {x} :
+  g x = to_add_comm_group f x :=
+@abelianization.to_comm_group.unique (free_group α) _ (multiplicative β) _ _ _ g
+  ⟨λ x y, @is_add_group_hom.add (additive $ abelianization (free_group α)) _ _ _ _ _ x y⟩ (λ x,
+  @free_group.to_group.unique α (multiplicative β) _ _ (g ∘ abelianization.of)
+    ⟨λ m n, is_add_group_hom.add g (abelianization.of m) (abelianization.of n)⟩ hg _) _
+
+theorem to_add_comm_group.ext
+  (g h : free_abelian_group α → β)
+  [is_add_group_hom g] [is_add_group_hom h]
+  (H : ∀ x, g (of x) = h (of x)) {x} :
+  g x = h x :=
+(to_add_comm_group.unique (g ∘ of) g (λ _, rfl)).trans $
+eq.symm $ to_add_comm_group.unique _ _ $ λ x, eq.symm $ H x
+

```

Comment:
Would it have been worth doing the last 40 or so lines in a `to_add_comm_group` namespace?

Reply:
added